### DISCUSSION TITLE
<del>我想引入一个验证库来验证从代码中拿到的 feature</del>

### DISCUSSION THREAD

**Discussion by timongh** [the1812/Bilibili-Evolved#3737] [功能建议 / Ideas]
现在的代码里，执行组件代码拿到 feature 后，是直接假装它是一个完整的 feature，没有任何验证。（代码写着都感觉不踏实）

我的倾向是，引入一个验证库，在使用之前验证其类型。初步选定的库是 Zod，次选 yup；这倒不急，我再看看。

当然，作为另一个选择，可以把从代码里拿到的东西标注为 `Partial<ComponentMetadata>` 之类的类型。之后每个实际用到的地方都要检测一下。

##### 更新

突然想到，似乎也没有必要验证。

反正发布出去的组件都是测试过的。用户自己安装的代码可以算作不在保护范围。如果用户自己修改了油猴里存的组件，也不在保护范围。

只要注意一下，不要让脚本因为这个错误一打开就崩溃就是了。

**Comment by the1812** [comment]
不会崩吧, 加载 feature 和执行 feature 都有 try / catch

**Reply by timongh** [reply]
这个，，具体没看。

不过对象就明明白白地放在那里，谁都能拿得到。说不定什么地方没注意就出错了呢。

比如要把组件显示在界面上，总得拿个 name 属性吧，这时候拿出来就是个 undefined。如果在什么不知道的地方对着  undefined 调了个方法，脚本就崩了。

如果脚本是在界面加载出来之前崩，那用户连删这个组件的机会都没有。

至于我提到的验证，你要是不嫌累赘的话，我可以等我闲的时候加上去。

**Reply by the1812** [reply]
这除非故意在运行时去改, 否则 name 不会是 undefined 的吧

**Reply by timongh** [reply]
不是在运行时改。我想要验证，是因为这个东西是外部输入的。既然是外部来的东西，那就什么情况都有可能。

不过我后面不是又说了嘛，可以假设输入功能代码都是完整且正确的。
